### PR TITLE
Add Searchfox link for metrics (fixes #65)

### LINKED ISSUE #65: Add searchfox links for metrics
It would be super handy to provide searchfox search URLs that search the code base for uses of a given metric.  It should be enough to search for `category_name.metric_name` as a string.

Depends on https://bugzilla.mozilla.org/show_bug.cgi?id=1668547

### PR DESCRIPTION
**Pull request by Iinh**
Fixes #65.

### Pull Request checklist

- [x] The pull request has a descriptive title (and a reference to an issue it
      fixes, if applicable)
- [x] All tests and linter checks are passing
- [x] The pull request is free of merge conflicts

### CODE REVIEW

File: src/components/SearchfoxLink.svelte [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+    "lockwise-android": "lockbox-android", // (?)
+
+    "lockwise-ios": "", // (?)
+
+    mach: "", // (?)
+
+    mozphab: "", // (?)
+
+    mozregression: "", // (?)
```
Comment (Iinh) [inline src/components/SearchfoxLink.svelte:51]: I was not able to find some applications in Searchfox. Could you help point me to where I might be able to find these repos @wlach? Thanks!

Comment (wlach) [reply src/components/SearchfoxLink.svelte:51]: Right. Someone needs to explicitly add a repo to be indexed to searchfox. I think we can just skip these repos for now, where they're not indexed -- we can always add them at a later time, and update this list. The major ones (e.g. Firefox for Android) are.

File: src/data/searchfox.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
+export const searchfoxMap = {
+  burnham: "",
```
Comment (wlach) [inline src/data/searchfox.js:2]: I would just omit entries from the map that aren't in searchfox.

File: src/pages/MetricDetail.svelte [hunk 3]
```diff
@@ -204,6 +207,14 @@
       <td>Version</td>
       <td>{metric.version || 0}</td>
     </tr>
+    {#if searchfoxMap[params.app]}
+      <tr>
+        <td>Find in</td>
```
Comment (wlach) [inline src/pages/MetricDetail.svelte:212]: Let's just use "Searchfox" here. It's on me to add some help text on what this means. 😄 

```suggestion
        <td>Searchfox</td>
```

File: src/pages/MetricDetail.svelte [hunk 4]
```diff
@@ -204,6 +207,14 @@
...
+        <td>
+          <SearchfoxLink metricName={metric.name} appName={params.app} />
```
Comment (wlach) [inline src/pages/MetricDetail.svelte:214]: Creating a component here seems like overkill to me. How about just extracting out the getSearchFoxLink method in the component and put into `src/formatters/searchfox.js`? Bonus: in that case you can skip importing `searchFoxMap` in this file, and just check if `getSearchFoxLink` returns a value.

**Review by wlach — CHANGES_REQUESTED** [review]
The ideas are right, but I think we can simplify this a bit.

File: src/pages/MetricDetail.svelte [hunk 5]
```diff
@@ -204,6 +206,14 @@
       <td>Version</td>
       <td>{metric.version || 0}</td>
     </tr>
+    {#if getSearchfoxLink(params.app, metric.name)}
+      <tr>
+        <td>Searchfox</td>
+        <td>
+          <a href={getSearchfoxLink(params.app, metric.name)}>{metric.name}</a>
+        </td>
```
Comment (wlach) [inline src/pages/MetricDetail.svelte:214]: I feel like maybe we should add some kind of warning if the metric is expired here, but we can do that in a followup.

**Review by wlach — CHANGES_REQUESTED** [review]
Some minor nits, but LGTM otherwise. The only other thing that concerns me a bit is that I'm actually not seeing many references to the metrics in the actual repositories. e.g. for for this fenix metric:

https://deploy-preview-339--glean-dictionary-dev.netlify.app/apps/fenix/metrics/activation-activation_id

We only get one result from the URL, a link to the generated markdown documentation:

https://searchfox.org/mozilla-mobile/search?q=activation.activation_id&path=fenix

In #65, @mdboom linked to https://bugzilla.mozilla.org/show_bug.cgi?id=1668547, but that landed quite a while ago so am a bit surprised not to see anything more come up for fenix. Is there more we need to do here?

@badboy @mdboom can you give more insight on this? Also I'd appreciate any thoughts you might have on the UI parts of this, I spend a lot less time in searchfox than either of you. 😄

**Comment by mdboom** [comment]
> @badboy @mdboom can you give more insight on this? Also I'd appreciate any thoughts you might have on the UI parts of this, I spend a lot less time in searchfox than either of you. smile

I think the crux of the problem is that the metric names are always in `snake_case` in the `metrics.yaml` file, but they may have different casing depending on the conventions of the various language.  For example, in Kotlin, members are in `camelCase`, so you'd have to search for [activation.activationId](https://searchfox.org/mozilla-mobile/search?q=activation.activationId&path=&case=false&regexp=false).  I think for the 80% solution, we'd need to generate a searchfox query with all of the possible casings, which are (where the first part is the metric category, and the second part the name):

- Kotlin & Swift: `CamelCase.camelCase`
- Python: `snake_case.snake_case`
- Rust: `snake_case::snake_case`

It's my bad thinking that https://bugzilla.mozilla.org/show_bug.cgi?id=1668547 would be helpful here -- that only affects generated code, which isn't available in searchfox.

File: src/formatters/searchfox.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  "firefox-ios-dev": "firefox-ios",
+
+  "firefox-ios-release": "firefox-ios",
+
+  "firefox-reality": "FirefoxReality",
+
+  "lockwise-android": "lockbox-android",
+
+  "reference-browser": "reference-browser",
+};
```
Comment (mdboom) [inline src/formatters/searchfox.js:27]: I wonder if (longer term -- maybe as a follow-on to this work) this should be in probe-scraper's repositories.yaml instead.

File: src/pages/MetricDetail.svelte (hunk 5 above)
Comment (mdboom) [reply src/pages/MetricDetail.svelte:214]: :+1:

**Comment by wlach** [comment]
> I think the crux of the problem is that the metric names are always in `snake_case` in the `metrics.yaml` file, but they may have different casing depending on the conventions of the various language. For example, in Kotlin, members are in `camelCase`, so you'd have to search for [activation.activationId](https://searchfox.org/mozilla-mobile/search?q=activation.activationId&path=&case=false&regexp=false). I think for the 80% solution, we'd need to generate a searchfox query with all of the possible casings, which are (where the first part is the metric category, and the second part the name):
> 
>     * Kotlin & Swift: `CamelCase.camelCase`
>     * Python: `snake_case.snake_case`
>     * Rust: `snake_case::snake_case`

Got it, I think this should not be too bad. @Iinh do you want to take a crack at this? Lodash's [capitalize](https://lodash.com/docs/4.17.15#capitalize) and [snakeCase](https://lodash.com/docs/4.17.15#snakeCase) functions may be useful here.

File: src/formatters/searchfox.js (hunk 6 above)
Comment (Iinh) [reply src/formatters/searchfox.js:29]: 👍

File: src/pages/MetricDetail.svelte (hunk 5 above)
Comment (Iinh) [reply src/pages/MetricDetail.svelte:224]: I'll add the warning in a followup PR after #341 has landed

**Comment by Iinh** [comment]
@wlach I added all the possible casings as @mdboom suggested, however I'm not sure if my approach to the Searchfox UI is correct. Guidance appreciated!

https://deploy-preview-341--glean-dictionary-dev.netlify.app/apps/fenix/metrics/activation_activation_id

Screen Shot 2021-01-25 at 11 49 16 AM

**Comment by mdboom** [comment]
> @wlach I added all the possible casings as @mdboom suggested, however I'm not sure if my approach to the Searchfox UI is correct. Guidance appreciated!
> 
> https://deploy-preview-341--glean-dictionary-dev.netlify.app/apps/fenix/metrics/activation_activation_id
> 
> Screen Shot 2021-01-25 at 11 49 16 AM

Personally, I'd prefer a single link that shows results for all of the languages (if that's possible with a searchfox query).  I don't think most users of this will know or care which language the telemetry is implemented in.  But that's just a hunch -- @chutten thoughts?

Also, it occurs to me that we'll need to support C++ and Javascript for Firefox desktop as well.  @chutten: would you mind providing a description of what those would look like [as in my comment above?](https://github.com/mozilla/glean-dictionary/pull/339#issuecomment-766904256).

Lastly, it would be great to have one of those informational tooltips on this field that says:

```
Finds uses of this metric using simple matching.
Metrics used dynamically may not appear in the results.
```

**Comment by wlach** [comment]
> Personally, I'd prefer a single link that shows results for all of the languages (if that's possible with a searchfox query). I don't think most users of this will know or care which language the telemetry is implemented in. But that's just a hunch -- @chutten thoughts?

searchfox supports regex queries, so we should be able to do this as follows:

https://searchfox.org/mozilla-mobile/search?q=(Activation.activationId|activation.activation_id)&path=fenix&regexp=true

> Lastly, it would be great to have one of those informational tooltips on this field that says:
> 
> ```
> Finds uses of this metric using simple matching.
> Metrics used dynamically may not appear in the results.
> ```

👍

**Comment by chutten** [comment]
C++ uses `snake_case::snake_case`
JS uses `camelCase.camelCase`

As for whether users would care what language it would be written in... definitely not (for Firefox Desktop). Firefox is written in JS and C++ and Rust and which language a probe is written in might be surprising if you don't understand the underlying patterns of which languages are commonly used for which subsystems. And then for different probes its language would only be surprising if you _did_ understand the patterns : |

**Comment by Iinh** [comment]
I added all the possible casings to Searchfox search query (Javascript, C++, Rust for `firefox-desktop`, and Kotlin, Swift, Python, Rust for mobile applications), thanks @mdboom, @wlach  and @chutten! However I notice that the search query for Glean metrics in [firefox-desktop](https://deploy-preview-339--glean-dictionary-dev.netlify.app/apps/firefox-desktop/) doesn't yield any result, e.g: 

https://deploy-preview-339--glean-dictionary-dev.netlify.app/apps/firefox-desktop/metrics/glean-error-io, or
https://deploy-preview-339--glean-dictionary-dev.netlify.app/apps/firefox-desktop/metrics/glean-error-preinit_tasks_overflow

Am I missing something?

**Comment by mdboom** [comment]
> I added all the possible casings to Searchfox search query (Javascript, C++, Rust for `firefox-desktop`, and Kotlin, Swift, Python, Rust for mobile applications), thanks @mdboom, @wlach and @chutten! However I notice that the search query for Glean metrics in [firefox-desktop](https://deploy-preview-339--glean-dictionary-dev.netlify.app/apps/firefox-desktop/) doesn't yield any result, e.g:
> 
> https://deploy-preview-339--glean-dictionary-dev.netlify.app/apps/firefox-desktop/metrics/glean-error-io, or
> https://deploy-preview-339--glean-dictionary-dev.netlify.app/apps/firefox-desktop/metrics/glean-error-preinit_tasks_overflow
> 
> Am I missing something?

In Glean, the app itself can define metrics, but also any libraries that the app uses (including Glean itself). In this case, these metrics are defined in Glean itself, so we'd need to search in Glean's github repo in searchfox.

As a first pass (to merge this PR), I'd say it's fine to only search the app for metrics (that's the common case).  As an additional hint, we could add to the tooltip "Only metrics defined in the application will be found."

To improve this, though, we'd need to search in the library the metric is defined in, not the app.  Does Glean dictionary have that information?  It's certainly known as part of the inputs to probe-scraper, but I'd have to check whether that's preserved in the output of probe-scraper...

**Comment by wlach** [comment]
> To improve this, though, we'd need to search in the library the metric is defined in, not the app. Does Glean dictionary have that information? It's certainly known as part of the inputs to probe-scraper, but I'd have to check whether that's preserved in the output of probe-scraper...

I believe it is not currently, but it's on my agenda to add this to a next-generation `/v2/` metrics API (proposal forthcoming).

**Comment by mdboom** [comment]
So, it seems like:

- For now, let's ignore the fact it can't find library metrics, and document that shortcoming in the tooltip.  We can file a bug and revisit this once the v2 probe-scraper format ships.
- Also, because of this multi-library thing, let's just search for all language patterns all the time (no reason to have different behavior for desktop). For example, even for Fenix (which is written in Kotlin), it uses libraries that are written in Rust.  I don't think there's a good reason to not just show all language patterns all the time.

**Review by mdboom — APPROVED** [review]
Looks good!  Let's not forget to file a follow-on issue to correctly support libraries when the other requirements for that exist.

**Review by wlach — APPROVED** [review]
💯

File: src/pages/MetricDetail.svelte (hunk 5 above)
Comment (Iinh) [reply src/pages/MetricDetail.svelte:220]: Should I still add a warning here @wlach, or do you think the expired Pill is sufficient enough?

Comment (wlach) [reply src/pages/MetricDetail.svelte:220]: I think it would be worth adding a quick warning. Doesn't hurt.

**Pull request merged**